pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,464 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Shot adaptive optimizer"""
+# pylint: disable=too-many-instance-attributes,too-many-arguments,too-many-branches
+from scipy.stats import multinomial
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+from .gradient_descent import GradientDescentOptimizer
+
+
+class ShotAdaptiveOptimizer(GradientDescentOptimizer):
+    r"""Optimizer where the shot rate is adaptively calculated using the variances of the parameter-shift
+    gradient.
+
+    By keeping a running average of the parameter-shift gradient and the *variance*
+    of the parameter-shift gradient, this optimizer frugally distributes a shot
+    budget across the partial derivatives of each parameter.
+
+    In addition, if computing the expectation value of a Hamiltonian using
+    :class:`~.ExpvalCost`, weighted random sampling can be used to further
+    distribute the shot budget across the local terms from which the Hamiltonian
+    is constructed.
+
+    .. note::
+
+        The shot adaptive optimizer only supports single QNodes or :class:`~.ExpvalCost` objects as
+        objective functions. The bound device must also be instantiated with a finite number
+        of shots.
+
+    Args:
+        min_shots (int): The minimum number of shots used to estimate the expectations
+            of each term in the Hamiltonian. Note that this must be larger than 2 for the variance
+            of the gradients to be computed.
+        mu (float): The running average constant :math:`\mu \in [0, 1]`. Used to control how quickly the
+            number of shots recommended for each gradient component changes.
+        b (float): Regularization bias. The bias should be kept small, but non-zero.
+        weighted_random_sampling (bool): Whether to use the weighted random sampling algorithm

```

Comment:
Might be good to mention how the shots are distributed when this is false.  Is it "Uniform deterministic sampling" (term according to Rosalin paper).

Maybe we could make this a string key, and later include the different types: uniform deterministic sampling, weighted deterministic sampling, weighted random sampling, and weighted hybrid sampling.

Reply:
Oh, this is a very good idea. I will do this.

Reply:
I've updated the argument to `term_sampling`, which can take two values for now:

- `weighted_random_sampling`
- None

This allows us to easily update it to support different sampling methods going forward.